tensorflow/compiler/xla/service/gpu/gpu_hlo_schedule.cc

```diff
@@ -184,6 +185,84 @@ void BFSLaunchOrder(const HloComputation* computation,
   }
 }
 
+bool CustomCallWithSchedule(const HloInstruction* instr,
+                            CustomCallSchedule schedule) {
+  return instr->opcode() == HloOpcode::kCustomCall &&
+         static_cast<const HloCustomCallInstruction*>(instr)
+                 ->custom_call_schedule() == schedule;
+}
+
+// Schedules EARLY_AS_POSSIBLE and LATE_AS_POSSIBLE custom-calls. This supports
+// a custom-call use case, where a logical operation is lowered into two HLOs
+// (e.g., PerformX and PerformXDone). We utilize this mechanism to either hide
+// host latencies between the pair of the custom-calls or more accurately
+// identify the def-use relationship of the two calls (typically PerformX is
+// scheduled right after all of its producers have been scheduled and
+// PerformXDone is scheduled right before its first consumer.)
+HloInstructionSequence postprocessor_to_custom_schedule(
+    const HloInstructionSequence& input) {
+  // Schedule `EARLY_AS_POSSIBLE`.
+  std::deque<HloInstruction*> early_as_possible_sched;
+  {
+    absl::flat_hash_set<HloInstruction*> scheduled;
+    const std::vector<HloInstruction*>& instrs = input.instructions();
+    for (HloInstruction* instr : instrs) {
+      if (scheduled.contains(instr)) {
+        continue;
+      }
+
+      early_as_possible_sched.push_back(instr);
+      scheduled.insert(instr);
+
+      for (HloInstruction* user : instr->users()) {
+        // Schedule any user who has the attribute `early_as_possible` and all
+        // of its producers have been scheduled.
+        if (CustomCallWithSchedule(user,
+                                   CustomCallSchedule::EARLY_AS_POSSIBLE) &&
+            absl::c_all_of(user->operands(), [&](const HloInstruction* opnd) {
+              return scheduled.contains(opnd);
```

Comment:
Should control-dependencies also be checked?

Reply:
Good question. I was also wondering if checking control dependencies is necessary in this particular use case of custom-calls. I think it might not be necessary as in the current code base, i.e., I don't find any codes adding control dependencies to custom-calls. However, let's be conservative and add the check for control dependencies for future proof.